metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/evaluator/BaselineRuleEvaluator.java

```diff
@@ -0,0 +1,194 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.math.stat.descriptive.moment.StandardDeviation;
+import org.apache.commons.math.stat.descriptive.rank.Median;
+import org.hypertrace.alert.engine.eventcondition.config.service.v1.DynamicThresholdCondition;
+import org.hypertrace.alert.engine.eventcondition.config.service.v1.MetricAnomalyEventCondition;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.AlertTask;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.DynamicRuleViolationSummary;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.EventRecord;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.MetricAnomalyNotificationEvent;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.NotificationEvent;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.ViolationSummary;
+import org.hypertrace.core.query.service.api.QueryRequest;
+import org.hypertrace.core.query.service.api.ResultSetChunk;
+import org.hypertrace.core.query.service.api.Row;
+import org.hypertrace.gateway.service.v1.baseline.Baseline;
+import org.hypertrace.gateway.service.v1.common.Value;
+import org.hypertrace.gateway.service.v1.common.ValueType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BaselineRuleEvaluator {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(BaselineRuleEvaluator.class);
+
+  private final QueryRequestHandler queryRequestHandler;
+
+  public BaselineRuleEvaluator(QueryRequestHandler queryRequestHandler) {
+    this.queryRequestHandler = queryRequestHandler;
+  }
+
+  Optional<NotificationEvent> evaluateRule(
+      MetricAnomalyEventCondition metricAnomalyEventCondition, AlertTask alertTask)
+      throws IOException {
+    DynamicThresholdCondition dynamicThresholdCondition =
+        metricAnomalyEventCondition
+            .getViolationConditionList()
+            .get(0)
+            .getDynamicThresholdCondition();
+
+    Duration duration = java.time.Duration.parse(dynamicThresholdCondition.getBaselineDuration());
+    long durationMillis = duration.toMillis();
+
+    // this query will fetch metric data, which will used for baseline calculation and evaluation
+    QueryRequest queryRequest =
+        queryRequestHandler.getQueryRequest(
+            metricAnomalyEventCondition,
+            alertTask.getTenantId(),
+            alertTask.getCurrentExecutionTime() - durationMillis,
+            alertTask.getCurrentExecutionTime());
+
+    LOGGER.debug("Rule id {}, Query request {}", alertTask.getEventConditionId(), queryRequest);
+
+    Iterator<ResultSetChunk> iterator =
+        queryRequestHandler.executeQuery(
+            Map.of(TENANT_ID_KEY, alertTask.getTenantId()), queryRequest);
+
+    List<Double> metricValuesForBaseline = new ArrayList<>();
+    List<Double> metricValuesForEvaluation = new ArrayList<>();
+
+    iterator.forEachRemaining(
+        chunk -> {
```

Comment:
Curious Q: Do we have to take care of the header of `ResultSetChunk`?

Reply:
Currently I do not see the need for using it